### PR TITLE
feat(ForcedDecisions): add forced-decisions APIs to OptimizelyUserContext

### PR DESCRIPTION
**Pull request by yasirfolio3**
## Summary
Add a set of new APIs for forced-decisions to OptimizelyUserContext:
- SetForcedDecision
- GetForcedDecision
- RemoveForcedDecision
- RemoveAllForcedDecisions

The following legacy APIs for feature management are deprecated in favour of `decide` APIs:
- IsFeatureEnabled
- GetFeatureVariable{Type}
- GetAllFeatureVariables
- GetAllFeatureVariablesWithDecision
- GetEnabledFeatures

## Test plan
- unit tests for the new APIs
- FSC tests with new test cases

### CODE REVIEW

**Review by msohailhussain — COMMENTED** [review]
partial review.

File: pkg/client/client.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ import (
 	"github.com/optimizely/go-sdk/pkg/config"
 	"github.com/optimizely/go-sdk/pkg/decide"
 	"github.com/optimizely/go-sdk/pkg/decision"
+	pkgReasons "github.com/optimizely/go-sdk/pkg/decision/reasons"
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/client/client.go:31]: This is because there are several properties named `reasons` throughout the code.

File: pkg/client/client.go [hunk 2]
```diff
@@ -52,7 +53,7 @@ type OptimizelyClient struct {
 // CreateUserContext creates a context of the user for which decision APIs will be called.
 // A user context will be created successfully even when the SDK is not fully configured yet.
 func (o *OptimizelyClient) CreateUserContext(userID string, attributes map[string]interface{}) OptimizelyUserContext {
-	return newOptimizelyUserContext(o, userID, attributes)
+	return newOptimizelyUserContext(o, userID, attributes, nil)
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/client/client.go:56]: Since we don't have any instance of forcedDecisionService to pass since a new `userContext` is being created. In this case, `newOptimizelyUserContext` will create a `forcedDecisionService` internally by itself .

File: pkg/config/datafileprojectconfig/mappers/forced_decision.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package mappers ...
+package mappers
+
+import (
+	datafileprojectconfig "github.com/optimizely/go-sdk/pkg/config/datafileprojectconfig/entities"
+	"github.com/optimizely/go-sdk/pkg/entities"
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/config/datafileprojectconfig/mappers/forced_decision.go:22]: since we are importing two `entities` files here, one has not be named.

File: pkg/client/optimizely_user_context.go [hunk 4]
```diff
@@ -29,23 +30,27 @@ type OptimizelyUserContext struct {
...
 
 // returns an instance of the optimizely user context.
-func newOptimizelyUserContext(optimizely *OptimizelyClient, userID string, attributes map[string]interface{}) OptimizelyUserContext {
+func newOptimizelyUserContext(optimizely *OptimizelyClient, userID string, attributes map[string]interface{}, forcedDecisionService *decision.ForcedDecisionService) OptimizelyUserContext {
 	// store a copy of the provided attributes so it isn't affected by changes made afterwards.
 	if attributes == nil {
 		attributes = map[string]interface{}{}
 	}
+	if forcedDecisionService == nil {
+		forcedDecisionService = decision.NewForcedDecisionService(userID)
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/client/optimizely_user_context.go:45]: `userID` is required here for logging purposes.

File: pkg/decision/forced_decision_service.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+type forcedDecision struct {
+	flagKey string
+	ruleKey string
+}
+
+// ForcedDecisionService defines user contexts that the SDK will use to make decisions for.
+type ForcedDecisionService struct {
+	UserID          string
+	forcedDecisions map[forcedDecision]string
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/decision/forced_decision_service.go:37]: `Structs` can be used as keys to a map if they hold comparable values which in our case is true since we have string values inside `forcedDecision `. https://go.dev/blog/maps 
I have benchmarked both map of maps and the above approach. Both approaches take almost equal time to read and write.

[the remark this answers is not in the document]
Comment (The-inside-man) [reply pkg/decision/forced_decision_service.go:37]: @msohailhussain Since go allows for the use of structs as keys then I think this is fine. My assumption is that the struct becomes some hash value in the background and that is what allows for quick lookups, same as if it were a single value. The only thing we would need to be careful of is if later we extend the forced decision struct to contain more data, as if we then update that object, the key will be different and take more than one place in the map if we are not careful and removing the older instance. This can also lead to a memory leak possibly if we dont clean up old objects that "get lost in the map". I am not 100% sure the Go implementation of this, but in Java that is an issue and why it should be used very cautiously. @jaeopt Do we see any extension of the forcedDecisions in the future? If so I would say avoid this method, otherwise - it should be alright for now.

[the remark this answers is not in the document]
Comment (msohailhussain) [reply pkg/decision/forced_decision_service.go:37]: agree !!!

[the remark this answers is not in the document]
Comment (msohailhussain) [reply pkg/decision/forced_decision_service.go:37]: OK I checked it and struct is much faster than conventional mapOfMap.
```
2021/09/29 11:50:43 mapOfMaps took 545.791µs
2021/09/29 11:50:43 mapOfForcedDecisions took 270.387µs
2021/09/29 11:50:43 mapOfMaps took 460.888µs
2021/09/29 11:50:43 mapOfForcedDecisions took 294.126µs
2021/09/29 11:50:43 mapOfMaps took 498.434µs
2021/09/29 11:50:43 mapOfForcedDecisions took 240.871µs
```

[the remark this answers is not in the document]
Comment (msohailhussain) [reply pkg/decision/forced_decision_service.go:37]: Insertion is bit slower in struct keys. 
```
2021/09/29 12:20:50 mapOfMaps took 914.17801ms
2021/09/29 12:20:53 forcedDecision took 2.856786962s
```

[the remark this answers is not in the document]
Comment (jaeopt) [reply pkg/decision/forced_decision_service.go:37]: It's not likely to extend FD in future more than what we have, but cannot be guaranteed. I think it'll be safe as long as we have good test cases. Performance-wise, I don't care much since this is for testing environment only :) @msohailhussain @The-inside-man

File: pkg/decision/forced_decision_service.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	return &ForcedDecisionService{
+		UserID:          userID,
+		forcedDecisions: map[forcedDecision]string{},
+		mutex:           new(sync.RWMutex),
+	}
+}
+
+// SetForcedDecision sets the forced decision (variation key) for a given flag and an optional rule.
+// returns true if the forced decision has been set successfully.
+func (f *ForcedDecisionService) SetForcedDecision(flagKey, ruleKey, variationKey string) bool {
```
Comment (jaeopt) [inline pkg/decision/forced_decision_service.go:52]: Need more comments about "empty" ruleKey case? Same for other APIs as well.

File: pkg/config/datafileprojectconfig/mappers/forced_decision.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+// MapFlagRules maps all rules (experiment rules and delivery rules) for each flag
+func MapFlagRules(featureFlags []datafileprojectconfig.FeatureFlag, experimentIDMap map[string]entities.Experiment, rolloutIDMap map[string]entities.Rollout) map[string][]entities.Experiment {
+	flagRulesMap := map[string][]entities.Experiment{}
+	experiments := []entities.Experiment{}
+	for _, flag := range featureFlags {
+		for _, experimentID := range flag.ExperimentIDs {
+			experiments = append(experiments, experimentIDMap[experimentID])
+			if rollout, ok := rolloutIDMap[flag.RolloutID]; ok {
+				experiments = append(experiments, rollout.Experiments...)
+			}
```
[the remark this answers is not in the document]
Comment (jaeopt) [reply pkg/config/datafileprojectconfig/mappers/forced_decision.go:34]: BTW, do we need "MapFlagRules"? I do not see any caller.

File: pkg/client/optimizely_user_context_test.go [hunk 8]
```diff
@@ -199,6 +199,40 @@ func (s *OptimizelyUserContextTestSuite) TestDecideFeatureTest() {
 	s.Equal("10418551353", impressionEvent.VariationID)
 }
 
+func (s *OptimizelyUserContextTestSuite) TestDecideFeatureTestWithForcedDecision() {
```
Comment (jaeopt) [inline pkg/client/optimizely_user_context_test.go:202]: Can we have one more test for empty "ruleKey"?

Comment (jaeopt) [reply pkg/client/optimizely_user_context_test.go:202]: Also, can we validate decision notification as well?

File: pkg/decision/forced_decision_service_test.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+	s.True(s.forcedDecisionService.SetForcedDecision("1", "2", "3"))
+	forcedDecision = s.forcedDecisionService.GetForcedDecision("1", "2")
+	s.Equal("3", forcedDecision)
+
+	s.True(s.forcedDecisionService.SetForcedDecision("1", "2", ""))
+	forcedDecision = s.forcedDecisionService.GetForcedDecision("1", "2")
+	s.Equal("", forcedDecision)
+
+	s.True(s.forcedDecisionService.SetForcedDecision("", "2", "3"))
+	forcedDecision = s.forcedDecisionService.GetForcedDecision("", "2")
```
Comment (jaeopt) [inline pkg/decision/forced_decision_service_test.go:76]: Are we going to accept "" flagKey as valid inputs?  It should be consistent with doc later.
If we define valid "flagKey" required, we can check non-empty flagKey when setForcedDecision.

**Review by jaeopt — CHANGES_REQUESTED** [review]
It overall looks great. See some suggestions and clarifications.

File: pkg/client/client.go [hunk 10]
```diff
@@ -563,6 +581,7 @@ func (o *OptimizelyClient) GetFeatureVariable(featureKey, variableKey string, us
 }
 
 // GetAllFeatureVariablesWithDecision returns all the variables for a given feature along with the enabled state.
+// Deprecated: Use 'decide' methods of 'OptimizelyUserContext' instead.
 func (o *OptimizelyClient) GetAllFeatureVariablesWithDecision(featureKey string, userContext entities.UserContext) (enabled bool, variableMap map[string]interface{}, err error) {
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/client/client.go:585]: Yes it's a public API which was only made available for `go`.

File: pkg/client/optimizely_user_context.go [hunk 11]
```diff
@@ -29,23 +30,27 @@ type OptimizelyUserContext struct {
...
+	}
 	attributesCopy := copyUserAttributes(attributes)
-
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/client/optimizely_user_context.go:43]: `forcedDecisionService` is not available to the user since its private, should we still be making a copy of it?

File: pkg/config/datafileprojectconfig/mappers/forced_decision.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	datafileprojectconfig "github.com/optimizely/go-sdk/pkg/config/datafileprojectconfig/entities"
+	"github.com/optimizely/go-sdk/pkg/entities"
+)
+
+// MapFlagRules maps all rules (experiment rules and delivery rules) for each flag
+func MapFlagRules(featureFlags []datafileprojectconfig.FeatureFlag, experimentIDMap map[string]entities.Experiment, rolloutIDMap map[string]entities.Rollout) map[string][]entities.Experiment {
+	flagRulesMap := map[string][]entities.Experiment{}
+	experiments := []entities.Experiment{}
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/config/datafileprojectconfig/mappers/forced_decision.go:28]: Forgot to clean this up, this method is no longer used.

File: pkg/decision/feature_experiment_service.go [hunk 13]
```diff
@@ -45,6 +45,21 @@ func (f FeatureExperimentService) GetDecision(decisionContext FeatureDecisionCon
...
+		if decisionContext.ForcedDecisionService != nil {
+			forcedDecision, reasons, err := decisionContext.ForcedDecisionService.FindValidatedForcedDecision(decisionContext.ProjectConfig, feature.Key, featureExperiment.Key, options)
+			if err == nil {
+				featureDecision := FeatureDecision{
+					Experiment: featureExperiment,
+					Variation:  forcedDecision,
+					Source:     FeatureTest,
+				}
+
+				return featureDecision, reasons, nil
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/decision/feature_experiment_service.go:59]: nice catch 👍

File: pkg/client/optimizely_user_context.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaeopt) [reply pkg/client/optimizely_user_context.go:43]: Theoretically possible for users to set/remove forcedDecisions while previous decides are ongoing or in other threads, so forcedDecisions should not be shared between multiple copies of userContext.

File: pkg/decision/forced_decision_service_test.go (hunk 9 above)
Comment (yasirfolio3) [reply pkg/decision/forced_decision_service_test.go:76]: Updated `setForcedDecision` to only accept non-empty flagKey.

**Review by jaeopt — CHANGES_REQUESTED** [review]
All changes look great!
I still see copying forcedDecision missing. Also consider optimization for creating forced-decision-service on the first setForcedDecision call. If not used for forced decision, it should be nil for normal decision flows.

**Comment by yasirfolio3** [comment]
> All changes look great! I still see copying forcedDecision missing. Also consider optimization for creating forced-decision-service on the first setForcedDecision call. If not used for forced decision, it should be nil for normal decision flows.

I have made the suggested changes but this is leading to all forcedDecision API's returning false or empty string until `setForcedDecision` is called once. We might have to update FSC tests accordingly too as currently, one of the tests expects `remove_all_forced_decisions` to return true in every case.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good. A couple of changes suggested.

File: pkg/client/optimizely_user_context.go [hunk 14]
```diff
@@ -85,21 +89,21 @@ func (o *OptimizelyUserContext) SetAttribute(key string, value interface{}) {
 // all data required to deliver the flag or experiment.
 func (o *OptimizelyUserContext) Decide(key string, options []decide.OptimizelyDecideOptions) OptimizelyDecision {
 	// use a copy of the user context so that any changes to the original context are not reflected inside the decision
-	userContextCopy := newOptimizelyUserContext(o.GetOptimizely(), o.GetUserID(), o.GetUserAttributes(), o.forcedDecisionService)
+	userContextCopy := newOptimizelyUserContext(o.GetOptimizely(), o.GetUserID(), o.GetUserAttributes(), o.getForcedDecisionService())
```
[the remark this answers is not in the document]
Comment (jaeopt) [reply pkg/client/optimizely_user_context.go:92]: My bad. It looks good as is :)

File: pkg/client/optimizely_user_context.go [hunk 15]
```diff
@@ -147,6 +160,9 @@ func (o *OptimizelyUserContext) RemoveAllForcedDecisions() bool {
 		o.optimizely.logger.Error("Optimizely instance is not valid, failing removeForcedDecision call.", err)
 		return false
 	}
+	if o.forcedDecisionService == nil {
+		return false
```
[the remark this answers is not in the document]
Comment (jaeopt) [reply pkg/client/optimizely_user_context.go:164]: Let's change this to "true". Logically false, but it should not change behavior depending on implementation. It still can be considered removed all successfully.

**Review by jaeopt — APPROVED** [review]
LGTM
